scripts/elasticsearch_nodes.py

```diff
@@ -0,0 +1,107 @@
+"""
+from itertools import (
+    repeat,
+)
+import json
+import logging
+import re
+from time import (
+    sleep,
+)
+
+from botocore.exceptions import (
+    ClientError,
+)
+import jq
+
+from azul import (
+    config,
+    require,
+)
+from azul.deployment import (
+    aws,
+)
+from azul.logging import (
+    configure_script_logging,
+)
+
+log = logging.getLogger(__name__)
+
+# How many times to retry the request to AWS if an error is encountered.
+max_attempts = 10
+
+
+def main() -> None:
+    for _ in range(max_attempts):
+        try:
+            response = aws.cloudwatch.list_metrics(Namespace='AWS/ES',
+                                                   MetricName='CPUUtilization',
+                                                   RecentlyActive='PT3H',
+                                                   Dimensions=[
+                                                       {
+                                                           'Name': 'DomainName',
+                                                           'Value': config.es_domain
+                                                       }
+                                                   ])
+        except ClientError as e:
+            log.debug('Error: ' + e.response['Error']['Code'] + '. Retrying...')
+        else:
+            nodes = [
+                d['Value']
+                for m in response['Metrics']
+                for d in m['Dimensions']
+                if d['Name'] == 'NodeId'
+            ]
+            if len(nodes) > 0:
+                break
+            else:
+                log.debug('Error: No node IDs found in response.')
+        sleep(5)
+    else:
+        raise RuntimeError('Failed to obtain node IDs. Max attempts exhausted.')
+
+    # The list of nodes IDs is appended with dummy entries to ensure the length
+    # matches the number of placeholders in the template. If our list had less
+    # than the template, Terraform would throw an "Invalid index" error. Also,
+    # the dummy entries cannot be empty strings.
+    placeholder_count = node_placeholder_count()
+    require(placeholder_count > 0,
+            'Error: No node ID placeholders found in template file.')
+    nodes = sorted(nodes)
+    nodes += list(repeat('blank', placeholder_count - len(nodes)))
+
+    # As required by Terraform for an external data source, we output the data
+    # as a map of string values.
+    result = {
+        "nodes": ','.join(nodes)
+    }
+    print(json.dumps(result))
+
+
+def node_placeholder_count() -> int:
+    """
+    Return the number of unique ES node ID placeholders found in the cloudwatch
+    dashboard template file. (i.e. the largest ${local.nodes[0]} index value +1)
+    """
+    with open(config.cloudwatch_dashboard_template) as f:
+        body = json.load(f)
+
+    largest_index = -1
+    for s in jq.all('.. | select(type == "string")', body):
+        m = re.findall(r'\$\{local.nodes\[(\d+)\]\}', s)
```

Comment:
I saw your comments but I think we are overthinking this. The brittleness I wanted to avoid is the hard-coded number of nodes because that changes all the time. Hard-coding a path into the JSON is also brittle but in a good way since it would alert us of format changes to the JSON, when we'd need to rethink other sections of the code.

Remember, the goal is not to avoid failure at all cost. Strictness and brittleness are different things. A failure is a legitimate outcome of a program. It's not the same as a crash or a coredump. Strictness means failing when internal assumptions aren't satisfied. Brittleness is making unreasonable assumptions, or assumptions that are in constant need of adjustment.

Reply:
Please raise in PL so we can fine-tune the `jq` query.